Feat: formatting and typescript 4

mint-dewit opened the pull request:
This PR adds a shared Sofie formatting + eslint config as well as an upgrade to typescript 4

Commit:
chore: add shared lint + format config prototype

Commit:
chore: formatting

Commit:
chore: centralized build config

Commit:
chore: add underscore types

Commit:
chore: add centralised code standard package

Commit:
fix: build errors after formatting

Note: removes the .event property (not sure what that was used for)

Commit:
chore: fix lint + build + test

Julusian commented on package.json:
```diff
@@ -116,12 +118,27 @@
     "source"
   ],
   "dependencies": {
+    "prettier": "^2.1.2",
```
prettier should be a devDependency
tslib needs updating to ^2.0.3

Julusian replied:
tslib still needs updating

Julusian commented on package.json:
```diff
@@ -97,10 +101,8 @@
-    "tslint-config-standard": "^8.0.1",
     "typedoc": "^0.15.0",
-    "typescript": "~3.6.0"
+    "typescript": "4"
```
Why is this just "4"? Shouldnt it be "~4.0" (or "4.0.x")
Same question about eslint.

I dont know how those get interpreted in semver for updates

mint-dewit replied:
No reason, good catch

Julusian commented on package.json:
```diff
@@ -84,11 +84,15 @@
     "jest": "^24.1.0",
+    "lint-staged": "^10.4.0",
     "mkdirp": "^0.5.1",
     "node-license-validator": "^1.3.0",
```
dont need `node-license-validator` anymore

Julusian commented on src/hyperdeck.ts:
```diff
@@ -18,40 +18,43 @@ class QueuedCommand {
 	}
+
+	resolve(_res: any) {
+		//
```
perhaps these should throw an error? That will help with debugging if in the future they dont get overridden

Julusian commented on tsconfig.json:
```diff
@@ -1,6 +1,16 @@
 {
-	"extends": "./config/tsconfig.base",
```
you should delete the old tsconfig files in /config/

Julusian reviewed (changes requested):
Nothing significant, just a few small questions/bits to tidy :)

nytamin commented on tsconfig.build.json:
```diff
@@ -1,9 +1,13 @@
-}
\ No newline at end of file
+	"extends": "@sofie-automation/code-standard-preset/ts/tsconfig.lib",
+	"exclude": ["node_modules/**", "src/**/*spec.ts", "src/**/__tests__/*", "src/**/__mocks__/*"],
```
could the excludes perhaps be moved up into the preset-repo?

mint-dewit replied:
could... probably? cc @Julusian 
should... that's a different question I guess. it would imply an enforced file structure, which is not something we enforce today (more or less assumed but not enforced)

Julusian replied:
The includes/excludes are all relative to the config file path, but I guess it should work though.
I think having it in this repo is fine though.
Also I am about to suggest another change which will cause a conflict with pushing it upstream too

Julusian commented on tsconfig.json:
```diff
@@ -1,6 +1,16 @@
-	]
-}
\ No newline at end of file
+	"extends": "@sofie-automation/code-standard-preset/ts/tsconfig.lib",
```
this should extend "./tsconfig.build.json", so that compilerOptions can be defined once and be used by both versions of the file

Commit:
chore: remove old ts config + clean up package.json + extend config

Commit:
fix: throw error for missing promise

Julusian commented on tsconfig.build.json:
```diff
@@ -1,9 +1,13 @@
+		"paths": {
+			"*": ["./node_modules/*", "./src/types/*"],
+			"hyperdeck-connection": ["./src/index.ts"],
+			"types": ["node", "jest"]
```
should types be inside paths?

gundelsby commented on package.json:
```diff
@@ -31,7 +31,7 @@
     "build": "trash dist && yarn build:main",
     "build:main": "tsc -p tsconfig.build.json",
-    "lint": "tslint --project tsconfig.json --config tslint.json",
+    "lint": "eslint . --ext .ts --ignore-pattern dist",
```
I'd prefer if we had ignores in an [.eslintignore](https://eslint.org/docs/user-guide/configuring#eslintignore) rather than specify it in the script, just like we do for prettier

Julusian replied:
If it was more than just ignoring `dist` then sure, but with it being such a simple rule I think that a file will just add unnecessary noise to the root level of the repo

gundelsby reviewed (changes requested):
Apart from the one comment, this looks good to me after  :)

Commit:
chore(ci): skip linting on node 8 test

Commit:
chore: ts config targets es6 (for node 8)

Commit:
chore: add eslintignore

Commit:
chore: upgrade jest, fix audit

Commit:
chore: fix jest config

Commit:
chore: upgrade typedoc

Commit:
!chore: drop node 8 support

With the upgrade to TS 4 it's no longer possible to keep unit tests for node 8 running, hence we have to drop support for node 8.

sparkpunkd reviewed (approved):
Looks good and already thoroughly reviewed.

I do have some comments about implication of changes here and things I found using the preset wrt the sofie-code-standard-preset repo that I have turned into a PR: https://github.com/nrkno/tv-automation-sofie-code-standard-preset/pull/1

Commit:
chore: upgrade dependencies

Commit:
chore: update config for new code-standard-preset
